services/smtp.py

```diff
@@ -0,0 +1,131 @@
+            return False
+        self.host = get_config("services", "smtp", "host", default="mailhog")
+        self.port = get_config("services", "smtp", "port", default=1025)
+        self.username = get_config("services", "smtp", "username", default=None)
+        self.password = get_config("services", "smtp", "password", default=None)
+        self.ssl_context = ssl.create_default_context()
+        return True
+
+    def try_starttls(self):
+        # only necessary if SMTP server supports TLS and authentication,
+        # for example mailhog does not need these two steps
+        try:
+            SMTPService.connection.starttls(context=self.ssl_context)
+        except smtplib.SMTPNotSupportedError:
+            log.warning(
+                "Server does not support TLS, continuing initialization of SMTP connection",
+                extra=dict(
+                    host=self.host,
+                    port=self.port,
+                    username=self.username,
+                    password=self.password,
+                ),
+            )
+        except smtplib.SMTPResponseException as exc:
+            log.warning("Error doing STARTTLS command on SMTP", extra=self.extra_dict)
+            raise SMTPServiceError("Error doing STARTTLS command on SMTP") from exc
+
+    def try_login(self):
+        try:
+            SMTPService.connection.login(self.username, self.password)
+        except smtplib.SMTPNotSupportedError:
+            log.warning(
+                "Server does not support AUTH, continuing initialization of SMTP connection",
+                extra=self.extra_dict,
+            )
+        except smtplib.SMTPAuthenticationError as exc:
+            log.warning(
+                "SMTP server did not accept username/password combination",
+                extra=self.extra_dict,
+            )
+            raise SMTPServiceError(
+                "SMTP server did not accept username/password combination"
+            ) from exc
+
+    def make_connection(self):
+        try:
+            SMTPService.connection.connect(self.host, self.port)
+        except smtplib.SMTPConnectError as exc:
+            raise SMTPServiceError("Error starting connection for SMTPService") from exc
+        self.try_starttls()
+        if self.username and self.password:
+            self.try_login()
+
+    def __init__(self):
+        if not self._load_config():
+            log.warning("Unable to load SMTP config")
+            return
+        if SMTPService.connection is None:
+            try:
+                SMTPService.connection = smtplib.SMTP(
+                    host=self.host,
+                    port=self.port,
+                )
+
+            except smtplib.SMTPConnectError as exc:
+                raise SMTPServiceError(
+                    "Error starting connection for SMTPService"
+                ) from exc
+
+            self.try_starttls()
+
+            if self.username and self.password:
+                self.try_login()
```

Comment:
would be nice to get rid of this duplicated code. it's already in `make_connection()`, could we just use that here too? maybe add the `if SMTPService.connection is None` stuff inside `make_connection()` too

not a blocking comment, and sorry if it was like this before and i missed commenting on it!

Reply:
it was not there, I added these methods in one of my recent commits, the reason I'm not using make_connection here is that here we are creating the `SMTP` object, but in make_connection we're assuming that this step has already been done and we're just running `connect` again on that existing object